modules/packages/Sort.chpl

```diff
@@ -18,53 +18,321 @@
+  // Compare results of comparator.key(a) if is defined by user
+  if canResolveMethod(comparator, "key", a) && canResolveMethod(comparator, "key", b) {
+    return chpl_compare(comparator.key(a), comparator.key(b));
+  // Use comparator.compare(a, b) if is defined by user
+  } else if canResolveMethod(comparator, "compare", a, b) {
+    return comparator.compare(a ,b);
   } else {
-    if reverse then return a > b;
-    else return a < b;
+    compilerError("The comparator record requires a 'key(a)' or 'compare(a, b)' method");
   }
 }
 
 
 /*
+    Check if a comparator was passed and confirm that it will work, otherwise
+    throw a compile-time error.
+
+    :arg a: Sample data passed to confirm that comparator methods can resolve
+    :arg comparator: Record that redefines the comparison mechanism with one of
+      the methods: ``comparator.key(a)`` or ``comparator.compare(a,b)``
+
+ */
+private proc chpl_check_comparator(comparator, type eltType) {
+  use Reflection;
+
+  // Dummy data for checking method resolution
+  // This may need updating when constructors support non-default args
+  const data: eltType;
+
+  if comparator.type == DefaultComparator {}
+  // Check for valid comaparator methods
+  else if canResolveMethod(comparator, "compare", data, data) {
+    // Check return type of compare
+    type comparetype = comparator.compare(data, data).type;
+    if !(isNumericType(comparetype)) {
+      compilerError("The compare method must return a numeric type");
+    }
+  }
+  else if canResolveMethod(comparator, "key", data) {
+    // Check return type of key
+    type keytype = comparator.key(data).type;
+    if !(isNumericType(keytype) || isStringType(keytype)) {
+      compilerError("The key method must return a numeric or string type");
+    }
+  }
+  else {
+    // If we make it this far, the passed comparator was defined incorrectly
+    compilerError("The comparator record requires a 'key(a)' or 'compare(a, b)' method");
+  }
+}
+
+
+/* Basic Functions */
+
+/*
+   Convenience wrapper around :proc:`quickSort`.
+
+   :arg Data: The array to be sorted
+   :type Data: [] `eltType`
+   :arg comparator: Record that redefines the comparison mechanism with one of
+      the methods: ``comparator.key(a)`` or ``comparator.compare(a,b)``
+
+ */
+proc sort(Data: [?Dom] ?eltType, comparator:?rec=defaultcomparator) where Dom.rank == 1 {
+  quickSort(Data, comparator=comparator);
+}
+
+
+/*
+   Check if array `Data` is in sorted order
+
+   :arg Data: The array to verify
+   :type Data: [] `eltType`
+   :arg comparator: Record that redefines the comparison mechanism with one of
+      the methods: ``comparator.key(a)`` or ``comparator.compare(a,b)``
+   :returns: ``true`` if array is sorted
+   :rtype: `bool`
+ */
+proc isSorted(Data: [?Dom] ?eltType, comparator:?rec=defaultcomparator): bool {
+  chpl_check_comparator(comparator, eltType);
+  for i in Dom.low..Dom.high-1 do
```

Comment:
I think it would be fun to write this in parallel but it doesn't need to happen right now...